data/policies/trait.cas

```diff
@@ -10,8 +10,9 @@ resource foo inherits my_trait {
 	}
 }
 
+@derive([write], strategy=union)
```

Comment:
This line is correct but not required, right? If we remove the `@derive` annotation we should get the same CIL, correct? What this test is actually testing?

Reply:
This test is testing #51.  Since derive support hadn't merged when we merged #51, the test only including explicitly implemented children, and this commit addresses a TODO there to test compatibility with the @derive feature.

The line is required.  The CIL with the derive annotation includes the line:

```
(macro baz-write ((type this) (type source)) (allow source this (file (write))))
```

which is generated by this derive.

If we remove the `@derive` annotation, it fails to compile with this error:

```
error: baz does not define a function named write
   ┌─ data/policies/trait.cas:13:23
   │
13 │ resource baz inherits my_trait {}
   │                       ^^^^^^^^ All types inheriting my_trait are required to implement write because it is marked as virtual

Error: Custom { kind: InvalidData, error: "Invalid policy" }
```

Which I now see has a wrong error message, since "is marked as virtual" is inaccurate.  It should be an error, but the message should say something about a trait instead.  I'll look into that, I thought I remembered fixing it a couple of weeks ago.  Maybe it's on another branch...

Regardless of the bad message, my_trait is a trait, which requires the implementation of the child function write, and this `@derive` is how baz is doing that.

Reply:
Ok, but I though because the `my_trait` has a default implementation it would not be required for `baz` to explicitly add a derive annotation. Do we really want to require such explicit annotation for a default implementation? I would not expect so, as well as if `baz` inherited `foo`.

Reply:
I think that the idea behind the default implementation was that it enables deriving.  Without such a default implementation you couldn't derive at all.

I think explicit is better when we're talking about security.  The hope with the `@derive` annotation is to allow developers ease of use without taking away control and having too many surprising side effects under the hood.  I think its consistent with rust and works well (some rust traits are derivable, and it's straightforward to derive them, but you can also provide your own implementation).

The idea here is that with inheriting my_trait, you are promising that you will provide some `write()` function matching the prototype.  It's up to the developer whether they want to just take the default, provide their own, or perhaps union several versions together with multiple inheritance, but we do require them to be explicit about it.  That way it's a conscious decision and if they don't think about it the compiler doesn't just decide to do something they may not expect, it prompts them to make a choice.

Also, as you mention with multiple inheritance, a default is really non-intuitive.  If `baz` inherited both `foo` and `my_trait`, does the fact that `my_trait` is a trait mean that the default should just be `my_trait`'s write?  Or should the default include a union with `foo` in that case?  I feel like people might want either, so it's better to be explicit.